(feat): add deployment page (#350)
chalk-ai/docs: deployment.mdx

published: true
---

import { ServiceDiagramSwitcher, CloudProviderAuth } from '@/components/ArchitectureDiagram'

---

## Overview

Chalk consists of two main components:
- A **Data Plane**, which hosts customer compute, and
- A **Metadata Plane**, which manages deployments to the Data Plane and other orchestration responsibilities.

> Typically, the Chalk team deploys the Data Plane into new cloud accounts that are owned and controlled by the customer. The Metadata Plane is owned and centrally managed by Chalk: it retains no customer application data, and is responsible for orchestration and configuration management.

Author: ```suggestion
Typically, the Chalk team deploys the Data Plane into new cloud accounts that are owned and controlled by the customer while the Metadata Plane is owned and centrally managed by Chalk. The Metadata plane retains no customer application data, and is responsible only for orchestration and configuration management.
```

> Typically, the Chalk team deploys the Data Plane into new cloud accounts that are owned and controlled by the customer while the Metadata Plane is owned and centrally managed by Chalk. The Metadata plane retains no customer application data, and is responsible only for orchestration and configuration management.

Author: ```suggestion
Typically, the Chalk team deploys the Data Plane into new cloud accounts that are owned and controlled by the customer while the Metadata Plane is owned and centrally managed by Chalk. The Metadata plane retains no customer application data and is responsible only for orchestration and configuration management.
```

## Data Plane Deployment

<CloudProviderAuth />

## Data Plane Structure

The Chalk Data Plane consists of a few primary components:
- **A Kubernetes Cluster**: hosts workloads, handles autoscaling and manages deployments.
- **A Metrics Database**: stores and aggregates feature and metadata metrics.
- **A Persistence Pipeline**: handles queuing for high frequency results and metrics.

> - **An Online Store**: caches features with explicit staleness to accelerate feature retrieval. We provide a [number of different configuration options for the online store](#).

Author: ```suggestion
- **An Online Store**: caches features with explicit staleness to accelerate feature retrieval. We provide a [number of different configuration options for the online store](/docs/choosing-online-offline-stores#online-store-options).
```

> - **An Offline Store**: handles large datasets (both created and ingested by Chalk engines), and query logging. We provide a [number of different configuration options for the offline store](#).

Author: ```suggestion
- **An Offline Store**: handles large datasets (both created and ingested by Chalk engines), and query logging. We provide a [number of different configuration options for the offline store](/docs/choosing-online-offline-stores#offline-store-options).
```

- **Secrets**: data plane secrets are hosted in the same cloud account as other resources, ensuring sensitive data stays confined to the Data Plane.
- **A Load Balancer**: routes traffic to various deployments.
- **Engines**: serve online and offline queries.

After deploying the Data Plane, we continue to monitor and update the system to ensure that engines are healthy, secure, and are serving traffic effectively.

## Post Deployment

> Once a Chalk Data Plane has been deployed, VPC peering is commonly added to allow private traffic between data sources in other accounts and the Data Plane. Additionally, some offline stores require additional integration steps and permissioning; we cover this in more data in [our offline store documentation](#). Adding additional data sources [is fully self-serve](#).

Author: ```suggestion
Once a Chalk Data Plane has been deployed, VPC peering is commonly added to allow private traffic between data sources in other accounts and the Data Plane. Additionally, some offline stores require additional integration steps and permissioning. Adding additional data sources [is fully self-serve](/docs/integrations).
```

## Security Considerations

The Chalk Metadata Plane can be restricted to have write-only access to secrets in the Data Plane. Conversely, the Data Plane can be given read-only access to secrets to enforce secure injection of secrets by users.

## Ongoing Permissions

> Chalk’s support team will work with you to scope permissions down to the subset required for ongoing maintenance. The precise details of this process depends on the level of ongoing support that your team needs and your compliance requirements. In principle, Chalk does not require ongoing access to data or the ability to edit IAM permissions: Chalk requires ongoing access to update the software deployed in your environment.

Author: ```suggestion
Chalk’s support team will work with you to scope permissions down to the subset required for ongoing maintenance. The precise details of this process depends on the level of ongoing support that your team needs and your compliance requirements. In principle, Chalk does not require ongoing access to data or the ability to edit IAM permissions. Chalk requires ongoing access to update the software deployed in your environment.
```
